### PR TITLE
Launch darkly implementation

### PR DESCRIPTION
**Pull request by nitishsingla1234**
### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/FeatureFlagToggle.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Inherited
+public @interface FeatureFlagToggle {
+    boolean flagEnabled() default true;
+}
```
[the remark this answers is not in the document]
Comment (nitishsingla1234) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/FeatureFlagToggle.java:14]: Moved the interface to Annotations folder.

File: src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentIntegrationTest.java [hunk 2]
```diff
@@ -40,6 +41,9 @@
     public static final String ROLE_ASSIGNMENT_ID = "2ef8ebf3-266e-45d3-a3b8-4ce1e5d93b9f";
     private MockMvc mockMvc;
 
+    @Rule
+    public FeatureFlagToggleEvaluator featureFlagToggleEvaluator = new FeatureFlagToggleEvaluator();
+
```
[the remark this answers is not in the document]
Comment (nitishsingla1234) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentIntegrationTest.java:46]: We are using Junit4 Rules and it's a standard syntax for initialising the rules.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/controller/endpoints/QueryAssignmentController.java [hunk 3]
```diff
@@ -81,8 +81,8 @@ public QueryAssignmentController(QueryRoleAssignmentOrchestrator queryRoleAssign
...
     }
 
-    @GetMapping(path = "/am/role-assignments/ld/endpoint1")
-    public ResponseEntity<Object> getIdLdDemo1() {
-        return ResponseEntity.status(HttpStatus.OK).body("Launch Darkly flag check is successful for endpoint 1");
+    @GetMapping(path = "/am/role-assignments/ld/endpoint")
+    public ResponseEntity<Object> getIdLdDemo() {
+        return ResponseEntity.status(HttpStatus.OK).body("Launch Darkly flag check is successful for the endpoint");
     }
 }
```
[the remark this answers is not in the document]
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/controller/endpoints/QueryAssignmentController.java:88]: This is a dummy endpoint. Once we have support for functional test cases, we will integrate LD with an actual endpoint.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/launchdarkly/FeatureConditionEvaluation.java [hunk 4]
```diff
@@ -1,42 +1,44 @@
...
                              @NotNull HttpServletResponse response, @NotNull Object arg2) throws Exception {
 
         Map<String, String> launchDarklyUrlMap = featureToggleService.getLaunchDarklyMap();
 
-        if (launchDarklyUrlMap.get(request.getRequestURI()).isEmpty()) {
-            response.sendError(403, "The endpoint is not configured in Launch Darkly");
+        if (launchDarklyUrlMap.get(request.getRequestURI()) == null
+            || launchDarklyUrlMap.get(request.getRequestURI()).isEmpty()) {
+            throw new ForbiddenException("The endpoint is not configured in Launch Darkly");
         }
```
[the remark this answers is not in the document]
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/launchdarkly/FeatureConditionEvaluation.java:34]: Yes, we will need to maintain a map between the feature flag name and the endpoint.
For example, if a user sends a request by postman, we don't have any way to get the flag name from the request(unless we add a header). Hence we need to maintain the flag at server side.

File: src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentIntegrationTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (shashank-rastogi) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentIntegrationTest.java:46]: Yea i understood.. my point was that .. does this need to be declared as a rule? if u r creating this as annotation then u can directly use annotation on the test.. is nt it? 
probably i am missing something but looks redundant to me

[the remark this answers is not in the document]
Comment (nitishsingla1234) [reply src/integrationTest/java/uk/gov/hmcts/reform/roleassignment/controller/RoleAssignmentIntegrationTest.java:46]: As discussed, we are using Junit 4 Test rule as Annotation processor. Once we move to Junit5, we will remove it.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/launchdarkly/FeatureConditionEvaluation.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/launchdarkly/FeatureConditionEvaluation.java:34]: We need to make an entry here only if an endpoint needs LD integration.

**Review by shashank-rastogi — APPROVED** [review]

**Pull request merged**